docs/patterns.md

```diff
@@ -77,3 +77,34 @@ module.export = (robot) ->
   robot.globalHttpOptions.httpAgent  = proxy('http://my-proxy-server.internal', false)
   robot.globalHttpOptions.httpsAgent = proxy('http://my-proxy-server.internal', true)
 ```
+
+## Dynamic matching of messages
+
+In some situations, you want to dynamically match different messages (e.g. factoids, JIRA projects). Rather than defining an overly broad regular expression that always matches, you can tell Hubot to only match when certain conditions are met.
+
+In a simple robot, this isn't much different from just putting the conditions in the Listener callback, but it makes a big difference when you are dealing with middleware: with the basic model, middleware will be executed for every match of the generic regex. With the dynamic matching model, middleware will only be executed when the dynamic conditions are matched.
+
+For example, the factoid lookup command could be reimplemented as:
+
+```coffeescript
+module.exports = (robot) ->
+  # Dynamically populated list of factoids
+  facts =
+    fact1: 'stuff'
+    fact2: 'other stuff'
+
+  robot.listen(
+    # Matcher
+    (message) ->
+      match = message.match(/^~(.*)$/)
+      # Only match if there is a matching factoid
+      if match and match[1] in facts
+        match[1]
+      else
+        false
+    # Callback
+    (response) ->
+      fact = response.match
```

Comment:
Does this work? I only see a `match` local variable in the matcher callback, nothing assigning it onto message object.